enhancements/sig-architecture/229-manifestworkreplicaset-rollback/README.md

```diff
@@ -0,0 +1,425 @@
+
+#### Story 2 — Manual rollback capability
+
+As a user, I want to manually roll back to a previous stable revision when needed. I should be able to update the ManifestWorkReplicaSet spec to reference a previous revision, and the controller should restore that revision across all clusters.
+
+### Implementation Details/Notes/Constraints
+
+The rollback mechanism leverages Kubernetes ControllerRevision resources to maintain a historical record of ManifestWorkReplicaSet changes. This approach is similar to how StatefulSet and DaemonSet controllers track revision history.
+
+Key implementation constraints:
+- Revision history is limited by `revisionHistoryLimit` to prevent unbounded storage growth
+- Automatic rollback (`abortOnFailure`) is opt-in and should be used with caution when CRDs are involved
+- Rollback operations require the previous revision to be available in ControllerRevision history
+- The controller uses hash-based collision detection to ensure unique revision naming
+
+### Risks and Mitigation
+
+1. **CRD Rollback Risk** - If the manifestwork template includes a CRD version update, rolling back the CRD resource can cause significant impact on the system and application. 
+   - **Mitigation**: Using the automatic rollback feature (`abortOnFailure`) should be opt-in and used with caution. Users should be aware of the implications when rolling back CRD changes.
+
+2. **Storage Growth Risk** - Maintaining revision history could lead to unbounded storage growth.
+   - **Mitigation**: The `revisionHistoryLimit` field allows users to control the maximum number of revisions retained. The controller will automatically prune old revisions.
+
+3. **Rollback Failure Risk** - If a rollback operation itself fails, the system could be left in an inconsistent state.
+   - **Mitigation**: The controller tracks rollback status through conditions and status fields, allowing for monitoring and manual intervention if needed.
+
+## Design Details
+
+The `abort` operation cancels the progressing rollout and rolls it back to the stable revision automatically without updating `.spec.manifestWorkTemplate`, whereas `rollback` requires explicit manual action by updating `.spec.manifestWorkTemplate`.
+
+### ManifestWorkReplicaSet API Object
+
+To support abort and rollback feature, we will have the following API changes:
+
+```yaml
+apiVersion: work.open-cluster-management.io/v1
+kind: ManifestWorkReplicaSet
+metadata:
+  name: <MWRS_Name>
+spec: 
+  # the number of revision history limit
+  revisionHistoryLimit: 3
+  placementRefs:
+    - name: placement-rollout-progressive
+      rolloutStrategy:
+        type: Progressive
+        progressive:
+          ...
+        # (NEW) abortOnFailure: if true, automatically aborts and rolls back on failure; otherwise, rollout ends on failure
+        abortOnFailure: true
```

Comment:
Will abort follow the same rolloutStrategy? Like auto rollback all at once, Progressive.

Reply:
In the first iteration, we will auto abort to all clusters for the simplicity.  Please let me know if you think we still need to follow rolloutStrategy for abort.

Reply:
I would define it as failureStrategy, with abortAll or rollbackAll as the strategy name. Try to avoid boolean type.